app/uk/gov/hmrc/emailverification/services/VerifiedEmailService.scala

```diff
@@ -0,0 +1,119 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.emailverification.services
+
+import config.{AppConfig, WhichToUse}
+import play.api.Logging
+
+import javax.inject.Inject
+import uk.gov.hmrc.emailverification.models.{MigrationResultCollector, VerifiedEmail}
+import uk.gov.hmrc.emailverification.repositories.{VerifiedEmailMongoRepository, VerifiedHashedEmailMongoRepository}
+
+import java.time.temporal.ChronoUnit
+import java.time.{Duration, Instant}
+import scala.concurrent.{ExecutionContext, Future}
+
+class VerifiedEmailService @Inject() (
+    verifiedEmailRepo:       VerifiedEmailMongoRepository,
+    verifiedHashedEmailRepo: VerifiedHashedEmailMongoRepository,
+    appConfig:               AppConfig
+)(implicit ec: ExecutionContext) extends Logging {
+
+  /**
+   * The older plain text emails collection was stored mixed case so there could be multiple entries of same email but different case.
+   * The newer hashed email collection lower cases emails before persisting.
+   */
+  def isVerified(mixedCaseEmail: String): Future[Boolean] = find(mixedCaseEmail).map(_.isDefined)
+
+  /**
+   * The older plain text emails collection was stored mixed case so there could be multiple entries of same email but different case.
+   * The newer hashed email collection lower cases emails before persisting.
+   */
+  def find(mixedCaseEmail: String): Future[Option[VerifiedEmail]] = appConfig.verifiedEmailCheckCollection match {
+    case WhichToUse.Old => verifiedEmailRepo.find(mixedCaseEmail)
+    case WhichToUse.Both => for {
+      oldVerifiedEmail <- verifiedEmailRepo.find(mixedCaseEmail)
+      newVerifiedEmail <- verifiedHashedEmailRepo.find(mixedCaseEmail.toLowerCase)
+    } yield {
+      if (oldVerifiedEmail.isDefined != newVerifiedEmail.isDefined) {
+        logger.warn(s"Email ${mixedCaseEmail.take(3)}***${mixedCaseEmail.takeRight(3)} only found in ${if (oldVerifiedEmail.isDefined) "old collection"}${if (newVerifiedEmail.isDefined) "new collection"} but expected in both.")
+      }
+      oldVerifiedEmail
+    }
+    case WhichToUse.New => verifiedHashedEmailRepo.find(mixedCaseEmail.toLowerCase)
+  }
+
+  /**
+   * older plain text collection is stored mixed case, new hashed collection is stored lower cased.
+   */
+  def insert(mixedCaseEmail: String): Future[Unit] = appConfig.verifiedEmailUpdateCollection match {
+    case WhichToUse.Both => verifiedHashedEmailRepo.insert(mixedCaseEmail.toLowerCase)
+      .flatMap(_ => verifiedEmailRepo.insert(mixedCaseEmail))
```

Comment:
I think we should insert first into the old repo, and if that succeeds, then insert into the new repo ... to keep them in-sync

Reply:
👍